### ISSUE TITLE
Show Level Length Instead of Tiny, Short, Medium, etc.

### ISSUE THREAD

**Issue by totallyphysicallypossible** [absoIute/Mega-Hack-Pro-Future#264]

**Comment by ExplodingBill** [comment]
it is possible to code

**Comment by Brittank88** [comment]
This is definitely possible but may show unavoidable inaccuracy (for example, if there are speed portals above the level, or you're intended to hit a speed portal from the top, bottom or front instead of the back side of the hitbox).

**Comment by NewAlexGamer** [comment]
simply displaying what the editor is saying could be good enough. if you want to be absolutely precise you'd need to first run a fuck ton of simulations inside the level to check if the taken path brings to death or completion. i suppose it would take too many resources to make it possible

**Comment by absoIute** [comment]
would require decoding the level string & calculating the level length, which would cause some lag (or could just be delayed if done on another thread), either way it's not a pretty implementation so not sure if i'll do this, but definitely will consider it as it has been requested a bit

**Comment by succmydecc** [comment]
i find it quite ironic that, his name is totallyphyiscallypossible but he says its probably not possible.

**Comment by Brittank88** [comment]
> would require decoding the level string & calculating the level length, which would cause some lag (or could just be delayed if done on another thread), either way it's not a pretty implementation so not sure if i'll do this, but definitely will consider it as it has been requested a bit

Could maybe 'pseudo-save' time by automatically calculating and storing lengths for any levels the player ever enters or opens in the editor? For any levels that aren't covered by this, I feel like off-thread would do the trick. Can just say "Calculating..." until it's done - that honestly seems pretty clean imo.

**Comment by succmydecc** [comment]
Well, instead of analyzing the entire level, how about analyzing the bit that shows you the time?

**Comment by Underkirby** [comment]
That could work

**Comment by Nicko2102** [comment]
> Well, instead of analyzing the entire level, how about analyzing the bit that shows you the time?

You still have to decode the whole thing i'm pretty sure

**Comment by Brittank88** [comment]
> Well, instead of analyzing the entire level, how about analyzing the bit that shows you the time?

Another thing he just invented out of thin air. I don't think level strings store the exact duration of the level, and you'd still have to decode the entire string to extract such information.

**Comment by Nicko2102** [comment]
> Another thing he just invented out of thin air. I don't think level strings store the exact duration of the level, and you'd still have to decode the entire string to extract such information.

that's what i was saying lol, the whole point of decoding it is that you have to do the whole string

**Comment by Brittank88** [comment]
> 

Which is why I suggested at least doing it when the string is already decoded.

**Comment by absoIute** [comment]
> 

Definitely could do this, 

> > would require decoding the level string & calculating the level length, which would cause some lag (or could just be delayed if done on another thread), either way it's not a pretty implementation so not sure if i'll do this, but definitely will consider it as it has been requested a bit
> 
> Could maybe 'pseudo-save' time by automatically calculating and storing lengths for any levels the player ever enters or opens in the editor? For any levels that aren't covered by this, I feel like off-thread would do the trick. Can just say "Calculating..." until it's done - that honestly seems pretty clean imo.

Definitely could do it this way, though considering just replacing this idea with a "time-left" label for in-game, I feel like that'd be more useful

**Comment by Brittank88** [comment]
> though considering just replacing this idea with a "time-left" label for in-game, I feel like that'd be more useful

I'd argue that both are useful. Knowing the full duration of a level could be good for speedruns and stuff. I've seen it before.

> Which is why I suggested at least doing it when the string is already decoded.

This is even easier than I thought when you enter the editor because you can directly rip the duration value that is already displayed lol.

**Comment by Underkirby** [comment]
That would be pretty easy to do as I think the level length is stored in
the level save file

**Comment by Nicko2102** [comment]
> That would be pretty easy to do as I think the level length is stored in the level save file

for that you would still have to decrypt it, whereas for brittank's method you just need to take the text that displays the length

**Comment by Underkirby** [comment]
That's what I meant

**Comment by Brittank88** [comment]
> 

To read the level save file in any useful way, you have to decrypt the entire string (it's all XOR'd, GZIP'd, and Base-64 encoded). I think this is what Nicko is getting at. Either way, yes, taking the length display text is probably easier than reading into level data.

**Comment by Underkirby** [comment]
I'm pretty sure that when you get sent the level data from the servers they are returned as an decrypted string

**Comment by Brittank88** [comment]
> I'm pretty sure that when you get sent the level data from the servers they are returned as an decrypted string

Actually, this could be true, I can't remember. Would potentially be a massive performance save for displaying the length of online levels in seconds. Depends on if that data is stored in the level string too.

**Comment by absoIute** [comment]
> This is even easier than I thought when you enter the editor because you can directly rip the duration value that is already displayed lol.

true but not something u can do for online levels rip

**Comment by absoIute** [comment]
> I'm pretty sure that when you get sent the level data from the servers they are returned as an decrypted string

Sadly they're not, you need to b64-decode & decompress

**Comment by Brittank88** [comment]
> > This is even easier than I thought when you enter the editor because you can directly rip the duration value that is already displayed lol.
> 
> true but not something u can do for online levels rip

Yeah, the amount of optimisation that can be applied depends on how you're viewing the level it seems. Maybe just run the decryption off-thread like you mentioned earlier? Seems like a fine solution to me. Just add a little loading circle and you're good tbh.
